tests/unittest_inference.py

```diff
@@ -4005,6 +4005,38 @@ class Test(Outer.Inner):
         assert isinstance(inferred, nodes.Const)
         assert inferred.value == 123
 
+    def test_infer_method_empty_body(self) -> None:
+        # https://github.com/PyCQA/astroid/issues/1015
+        node = extract_node(
+            """
+            class A:
+                def foo(self): ...
+
+            A().foo()  #@
+        """
+        )
+        inferred = next(node.infer())
+        assert isinstance(inferred, nodes.Const)
+        assert inferred.value is None
+
+    def test_infer_method_overload(self) -> None:
+        # https://github.com/PyCQA/astroid/issues/1015
+        node = extract_node(
+            """
+            class A:
+                def foo(self): ...
+
+                def foo(self):
+                    yield
+
+            A().foo()  #@
+        """
+        )
+        inferred = list(node.infer())
+        assert len(inferred) == 2
+        assert isinstance(inferred[0], nodes.Const)
+        assert isinstance(inferred[1], Generator)
```

Comment:
I'm wondering if
```py
class A:
    def foo(self): ...
    def foo(self):
        yield
```
should be inferred as `Const` and `Generator` as Python will always use the last definition.

--
Some test cases to be aware of
```py
if some_test:
    def func():
        return 42
else:
    def func():
        return bool

# should still return two results for `func()`
```

Functools.singledispatch should still work:
https://docs.python.org/3/library/functools.html#functools.singledispatch

Reply:
> Wouldn't it be better to only infer it as `Generator`?

My thinking leans the same way as it's the more "useful" definition.

> Python will always use the last definition

Pretty much, but I wonder what to do about monkeypatching, or this (contrived) example:

```python
class A:
    def foo(self):
        return 'first'
    bar = foo(object())
    def foo(self):
        return 'second'

A().foo()  # 'second'
A().bar  # 'first'
```

Reply:
Just tested your example: This will already be inferred incorrectly. `A().foo() ->'first'`. Possibly another bug?

As for the issue at hand. Would it work to filter previous definitions only if no other attributes are in between?